source/opt/reduce_load_size.cpp

```diff
@@ -0,0 +1,176 @@
+      composite_type->kind() == analysis::Type::kMatrix) {
+    return false;
+  }
+
+  ir::Instruction* var = composite_inst->GetBaseAddress();
+  if (var == nullptr || var->opcode() != SpvOpVariable) {
+    return false;
+  }
+
+  SpvStorageClass storage_class = static_cast<SpvStorageClass>(
+      var->GetSingleWordInOperand(kVariableStorageClassInIdx));
+  switch (storage_class) {
+    case SpvStorageClassUniform:
+    case SpvStorageClassUniformConstant:
+    case SpvStorageClassInput:
+      break;
+    default:
+      return false;
+  }
+
+  // Create a new access chain and load just after the old load.
+  // We cannot create the new access chain load in the position of the extract
+  // because the storage may have been written to in between.
+  InstructionBuilder ir_builder(inst->context(), composite_inst,
+                                ir::IRContext::kAnalysisInstrToBlockMapping |
+                                    ir::IRContext::kAnalysisDefUse);
+
+  uint32_t pointer_to_result_type_id =
+      type_mgr->FindPointerToType(inst->type_id(), storage_class);
+  assert(pointer_to_result_type_id != 0 &&
+         "We did not find the pointer type that we need.");
+
+  analysis::Integer int_type(32, false);
+  const analysis::Type* uint32_type = type_mgr->GetRegisteredType(&int_type);
+  std::vector<uint32_t> ids;
+  for (uint32_t i = 1; i < inst->NumInOperands(); ++i) {
+    uint32_t index = inst->GetSingleWordInOperand(i);
+    const analysis::Constant* index_const =
+        const_mgr->GetConstant(uint32_type, {index});
+    ids.push_back(const_mgr->GetDefiningInstruction(index_const)->result_id());
+  }
+
+  ir::Instruction* new_access_chain = ir_builder.AddAccessChain(
+      pointer_to_result_type_id,
+      composite_inst->GetSingleWordInOperand(kLoadPointerInIdx), ids);
+  ir::Instruction* new_laod =
+      ir_builder.AddLoad(inst->type_id(), new_access_chain->result_id());
+
+  context()->ReplaceAllUsesWith(inst->result_id(), new_laod->result_id());
+  context()->KillInst(inst);
+  return true;
+}
+
+bool ReduceLoadSize::ShouldReplaceExtract(ir::Instruction* inst) {
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  ir::Instruction* op_inst = def_use_mgr->GetDef(
+      inst->GetSingleWordInOperand(kExtractCompositeIdInIdx));
+
+  if (op_inst->opcode() != SpvOpLoad) {
+    return false;
+  }
+
+  auto cached_result = should_replace_cache_.find(op_inst->result_id());
+  if (cached_result != should_replace_cache_.end()) {
+    return cached_result->second;
+  }
+
+  bool all_elements_used = false;
+  std::set<uint32_t> elements_used;
+
+  def_use_mgr->ForEachUser(
+      op_inst, [&elements_used, &all_elements_used](ir::Instruction* use) {
+        if (use->opcode() != SpvOpCompositeExtract) {
+          all_elements_used = true;
+        }
+        elements_used.insert(use->GetSingleWordInOperand(1));
+      });
+
+  bool should_replace = false;
+  if (all_elements_used) {
+    should_replace = false;
```

Comment:
Should we have a threshold different than 100% at which to do the replacement?  On big structures, if more than half the elements are used, maybe the replacement is not worth doing?  Not necessary to do this in this PR.  I'm not even sure if it's worth implementing at all.

Reply:
In the other case, where not every element is used, we check if the percent used is less than a threshold.  That value is currently 90%.  We can adjust that as we see the need.